### ISSUE TITLE
We should make extensions more generic

### ISSUE THREAD

**Issue by westonpace** [substrait-io/substrait#162]
Right now we list all extensions used at the top of the plan.  However, we group items into three different categories of extensions, types, type variations, and functions.

So we have something like:

```
{
  "functions": ["custom_add", "magnitude"],
  "types": ["complex"],
  "type_variations": ["unsigned"]
}
```

However, these three categories are insufficient.  We already have examples of additional categories (e.g. file formats ala #138, `ReadRel::ExtensionTable`, `ExchangeRel::ExchangeTarget`, etc.) and it seems likely we will have more in the future (data serialization formats, metadata catalogs, etc.)

We could get rid of these categories entirely: https://github.com/westonpace/substrait/commit/a9ea46fd924d0381149bf93232bc8a4755887fd3

This gives us:

```
{ "extensions": [ "custom_add", "magnitude", "complex", "unsigned" ] }
```

We could also make "extension type" a first class concept with it's own fully qualified name: https://github.com/westonpace/substrait/commit/9882d04b6bd89ca015743a4895b0228c0e9cfdd0

This gives us:

```
{
  "extensions": [
    { "extension_type": "function", "name": "custom_add" },
    { "extension_type": "function", "name": "magnitude" },
    { "extension_type": "type", "name": "complex" },
    { "extension_type": "type_variation", "name": "unsigned" }
  ]
}
```

I don't have a strong preference on approach and am happy to consider alternatives.  Whatever we decide on I'm also happy to help write up the spec.  However, I expect this is going to have some backwards incompatible changes so I think it would be better to figure this out sooner rather than later.

**Comment by westonpace** [comment]
To help explain the need let me demonstrate a potential problem with the current approach.  Let's pretend we are creating a read of an extension table.  The data is in an existing SQL table in SQL server and we are going to use a JDBC connection to get it out (this is purely an example).

```
"read": {
  "base_schema": {...},
  "filter": {...},
  "projection": {...},
  "extension_table": {
    "detail": {
      "connection_string": "Server=myServerName\myInstanceName;Database=myDataBase;User Id=myUsername;Password=myPassword;"
    }
  }
}
```

The producer creates the above plan and sends it to the consumer.  The consumer sees:

```
  "extension_table": {
    "detail": { ... }
  }
```

The consumer now has no idea how to interpret "detail".  Maybe the consumer supports this use case.  Maybe they don't.  Maybe they support several different kinds of extension tables.  They have no idea of knowing what kind of extension table is being specified here.  On the other hand we can do:

```
  "extension_table": {
    "type": "jdbc",
    "detail": { ... }
  }
```

Now the consumer, assuming they understand and implement the "jdbc extension table" concept, can completely fulfill this plan.  It knows (based on some agreement between producer and consumer) exactly what the structure of `detail` should be when the extension table type is `jdbc`.

Of course, this is pretty much the same dilemma we face when mapping functions or types.  We will run into the same concerns of "Gee, it would be nice if we could associate this type with a URI" and "Huh, writing all these long URIs in our plan is somewhat inefficient".  So it seems to me that we would eventually arrive at (this is the second proposal in my earlier comment):

```
  "extension_table": {
    "extension_table_type_anchor": 73,
    "detail": { ... }
  }
  ...
  "extension_uris": [
    {
      "extension_uri_anchor": 10,
      "uri": "https://substrait.io/extensions/jdbc.yaml"
    }
  ]
  "extensions": [
    {
      "extension_uri_reference": 10,
      "extension_anchor": 73,
      "name": "jdbc_table"
    }
  ]
```

Of course, we can have fun deciding what kind of metadata we put in `jdbc.yaml` but that is besides the point for this PR.  We could accomplish our goals without that YAML file ever existing (just using the URI as a namespace).  That being said, I like the first proposal in my earlier comment because then we can pick a "YAML understander" based on the "extension type".

**Comment by jvanstraten** [comment]
I get the feeling that you're conflating protobuf `any` messages with just some binary blob of data. `any` messages do, in fact, store a fully qualified type name that should be globally unique (in addition to said binary blob of data). So, rather than seeing:

```
"extension_table": {
  "detail": { ... }
}
```

a consumer that doesn't understand the extension should still have the following information:

```
"extension_table": {
  "detail": {
    "@type": "some_hopefully_sufficiently_unique_namespace.JbdcTable",
    "@data": [...]
  }
}
```

I think that's the correct protobuf JSON syntax, but don't quote me on that. In C++, I believe the syntax to query the URL would be `detail().type_url()`, but Google's API documentation is pretty terrible so I could be wrong there, too. Syntax aside though, I know for a fact this information exists, because the Rust bindings I'm using for the validator give me precisely a type URL string and a binary string, and they round-trip.

Note that the type URL should already be "globally unique" on its own, because the protobuf library itself will throw a fit if two different generated classes with the same type URL are registered with its global state. I hold rather strong... let's say "opinions" about this mechanism, but that's how it works.

With that in mind, I'm not sure what problems your suggestions solve, while at the same time they add complexity and break backwards compatibility. You can certainly make a case for converting more extension concepts from "advanced" to "simple," but that can also be done in ways that don't change the message structure in a breaking way. You'd just add support for a new key at the root of the YAML file schema, extend the `extensions.SimpleExtensionDeclaration.mapping_type` oneof, and replace the advanced extension fields being replaced with a reference to the anchor (using a previously unused field index).

**Comment by westonpace** [comment]
You're correct.  I read more on this and hadn't realized that there was a unique URI already defined by protobuf for types.  So as long as our extensions are unique message types this should be sufficient.

Given this I would question what makes types & functions unique?  For example, why do we not define a custom data type with:

```
message Type {
  oneof kind {
    ...
    uint32 user_defined_type_reference = 31;
  }
  ...
}
```

When we could do this:

```
message Type {
  oneof kind {
    ...
    google.protobuf.Any user_defined_type = 31;
  }
  ...
}
```
I think if we can answer that question then we answer the question of whether we might want different kinds of extensions in the future.

**Comment by westonpace** [comment]
Alright, that works for me.
